jax/_src/lax/control_flow/loops.py

```diff
@@ -219,18 +223,40 @@ def scan(f, init, xs, length=None):
       length, = unique_lengths
 
   if config.jax_disable_jit:
-    if length == 0:
-      raise ValueError("zero-length scan is not supported in disable_jit() mode because the output type is unknown.")
-    carry = init
-    ys = []
-    maybe_reversed = reversed if reverse else lambda x: x
-    for i in maybe_reversed(range(length)):
-      xs_slice = [_index_array(i, core.get_aval(x), x) for x in xs_flat]
-      carry, y = f(carry, tree_unflatten(xs_tree, xs_slice))
-      ys.append(y)
-    stack = lambda *ys: jax.numpy.stack(ys)
-    stacked_y = tree_map(stack, *maybe_reversed(ys))
-    return carry, stacked_y
+    try:
+      if length == 0:
+        raise ValueError("zero-length scan is not supported in disable_jit() mode because the output type is unknown.")
+      carry = init
+      ys = []
+      maybe_reversed = reversed if reverse else lambda x: x
+      for i in maybe_reversed(range(length)):
+        if early_exit(carry):
+          break
+        else:
+          xs_slice = [_index_array(i, x) for x in xs_flat]
+          carry, y = f(carry, tree_unflatten(xs_tree, xs_slice))
+          ys.append(y[None])
+      if len(ys) != length:
+        if len(ys) == 0:
+          raise ValueError("zero-length scan (due to early exit) is not supported in "
+                           "disable_jit() mode because the output type is unknown.")
+        padding = tree_map(partial(_zeros_like, length - len(ys)), ys[0][0])
+        ys.append(padding)
+      concatenate = lambda *ys: jax.numpy.concatenate(ys)
+      stacked_y = tree_map(concatenate, *maybe_reversed(ys))
+      return carry, stacked_y
+    except core.ConcretizationTypeError:
+      # e.g. due to vmap or checkpoint. In this case fall back to primitive version
+      pass
+
+  def aug_f(carry, x):
+    out = f(carry, x)
+    if not hasattr(out, "__len__") or len(out) != 2:
+      msg = "scan body output must be a pair, got {}."
+      raise TypeError(msg.format(out))
+    carry, y = out
+    done = early_exit(carry)
+    return done, carry, y
```

Comment:
As you can see, the approach I took was to add the conditional as another output of the body, rather than as a separate jaxpr.

Upside: this really helped to minimise the amount of code changes needed elsewhere. No keeping track of separate constants for both `f` and `early_exit`, no extra fixed-point resolution needed, etc.

Downside: `early_exit` is traced/compiled twice; once at initialisation and once in the body.

Conditionals are generally small and cheap, so I felt this was a worthy trade-off here.